packages/input-phone/src/PhoneInput.js

```diff
@@ -0,0 +1,310 @@
+    [phone, selectedCountry, defaultCountry],
+  );
+  const filteredCountries = useCountrySearch(
+    countries,
+    countryQuery,
+    priorityCountries,
+  );
+
+  useEffect(() => {
+    if (!value) {
+      setSelectedCountry(null);
+    }
+  }, [value]);
+
+  const triggerChange = useCallback(
+    (number, code = country) => {
+      onChange?.(
+        parsePhoneNumberFromString(number, code)?.getURI() ||
+          `tel:+${getCountryCallingCode(code)}${number}`,
+      );
+    },
+    [country, onChange],
+  );
+
+  const handleNumberChange = useCallback(
+    ({ target }) => {
+      const newValue = parseDigits(target?.value);
+      triggerChange(newValue);
+    },
+    [country],
+  );
+
+  const handleCountrySelection = useCallback(code => {
+    setSelectedCountry(code);
+    setCountryQuery("");
+    triggerChange("", code);
+    dismissRef.current?.();
+  }, []);
+
+  const { handleKeyDown, highlightIndex, resetIndex } = useHighlight({
+    items: filteredCountries,
+    listRef,
+    selectFn: handleCountrySelection,
+    cancelFn: dismissRef.current,
+  });
+
+  const handleSearchChange = useCallback(event => {
+    setCountryQuery(event.target.value);
+    resetIndex();
+    listRef.current?.scrollTo(0);
+  }, []);
+
+  const handlePopoverOpen = useCallback(() => {
+    setPopoverOpen(true);
+    resetIndex();
+    setImmediate(() => {
+      searchInputRef.current?.focus?.();
+    });
+  }, []);
+  const handlePopoverClose = useCallback(() => {
+    setPopoverOpen(false);
+    setImmediate(() => {
+      numberInputRef.current?.focus?.();
+    });
+  }, []);
+
+  const listData = useMemo(
+    () => ({
+      items: filteredCountries,
+      handler: handleCountrySelection,
+      highlightIndex,
+    }),
+    [filteredCountries, highlightIndex],
+  );
+
+  return (
+    <TextInput
+      innerRef={numberInputRef}
+      prefix={
+        <Popover
+          onOpen={handlePopoverOpen}
+          onClose={handlePopoverClose}
+          tooltipProps={{
+            padding: "0",
+            fullScreenBreakpoint,
+            ...tooltipProps,
+          }}
+          disabled={disabled}
+          trigger={
+            <CountrySelectorTrigger
+              disabled={disabled}
+              callingCode={getCountryCallingCode(country)}
+              flag={flags[country]}
+              code={country}
+            />
+          }
+          content={dismiss => {
+            dismissRef.current = dismiss;
+            return (
+              // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
+              <div
+                onClick={event => {
+                  // We stop propagation to avoid giving focus to the main input
+                  // This happens because the popover is inside the input wrapper
+                  event.stopPropagation();
+                }}
+                css="
+                display: flex;
```

Comment:
I think it will be good to extract this and the `ul` bellow to styled-components.
PhoneInput is already heavily loaded, to improve this maybe you can decompose all the `content` render to another component or create a container to handle all these logic, which will make a good lean.

Reply:
I tried extracting the whole popover content, but it needed so many props that it made it worse. I ended up extracting just the `ul`.